ci: run s3tests

m-ildefons opened the pull request:
Run s3 test suite

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

m-ildefons committed:
ci: run s3tests

Run s3 test suite

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

jecluis commented on .github/workflows/build-test-radosgw.yml:
```diff
@@ -43,6 +47,25 @@ jobs:
+            ceph/build/bin/radosgw \
+            ceph/build/lib/libceph-common.so \
+            ceph/build/lib/libceph-common.so.2 \
+            ceph/build/lib/libradosgw.so \
```
haven't the `libradosgw.so*` files been removed?

jecluis reviewed (changes requested):
looks good except for the `libradosgw.so` files being packaged.

m-ildefons committed:
ci: build/test workflow remove libradosgw

Remove libradosgw from build/test workflow artifacts as it is no longer
being created.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
ci: s3tests migrate workflow to use s3tests runner

Migrate the workflow from using ad-hoc code to using the s3tests-runner
script. This script incorporates sensible log and result collection in
preparation for automatic archival of said results.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

irq0 commented on .github/workflows/build-test-radosgw.yml:
```diff
@@ -42,6 +46,22 @@ jobs:
+            ceph/build/lib/librados.so.2 \
+            ceph/build/lib/librados.so.2.0.0
+
+      - name: Save radosgw binary
```
Nit: plural or artifacts

m-ildefons replied:
Not sure what you mean?

irq0 replied:
name doesn't do what it says

irq0 commented on .github/workflows/build-test-radosgw.yml:
```diff
@@ -113,3 +133,98 @@ jobs:
+      matrix:
+        op:
+          - create
+          # - list
```
Nit: comment on why this is commented out would be nice

m-ildefons replied:
I removed the build matrix for other reasons. This was commented out due to me not yet having gone through the listing tests and curated them by passing/failing.

irq0 commented on .github/workflows/build-test-radosgw.yml:
```diff
@@ -113,3 +133,98 @@ jobs:
+            python3-nose \
+            wget
+
+          # Unfortunately, since yq is only available through snap on Ubunut and
```
Ubuntu

m-ildefons replied:
Ack.

irq0 commented on .github/workflows/build-test-radosgw.yml:
```diff
@@ -113,3 +133,98 @@ jobs:
+          export S3TEST_CONF=$FIXTURES/s3tests.conf
+
+          pushd s3tests
+          for t in $(yq \
```
the yq dependency seems more trouble than its worth, especially with the rather simple 'get map key, output as space sep list' usage.

Consider python instead:

```
python <<EOF
from yaml import load
print(" ".join(load(open('s3-tests.yaml'))["bucket_create"]))
EOF

```

use either with for t in ..
or better use `\n` and pipe to while read t; do ..

jecluis replied:
this also presumes 'PyYaml' as an installed dependency though. I agree with not installing "random" files from the internet, but doing it with Python will require the dependency to be installed.

m-ildefons replied:
`yq` is needed for the s3tests-runner script anyways, as it is what I use to compile an aggregated report about which tests pass and which fail. I'm not keen on re-implementing what I need from it in python, especially not the more complex parts like inserting into maps and converting to JSON. That being said, I pinned the version down with a checksum check for good measure.

irq0 replied:
:+1:

irq0 commented on .github/workflows/build-test-radosgw.yml:
```diff
@@ -113,3 +133,98 @@ jobs:
+          # that doesn't work in docker containers (at least not out of the
+          # box), this abomination is the way to go to install yq.
+          sudo wget \
+            "${YQ_GH_URL}/releases/latest/download/yq_linux_amd64" \
```
This is basically code execution in our build workers by the yq project.
latest may point to anything. For basic security we may want to pin to version and do a checksum check. Better: replace with python (see below)

m-ildefons replied:
Ack. See above

irq0 commented on .github/workflows/build-test-radosgw.yml:
```diff
@@ -113,3 +133,98 @@ jobs:
+        run: |
+          set -x
+
+          export CEPH_DIR=${GITHUB_WORKSPACE}/ceph
```
Needs quoting. This line and below

m-ildefons replied:
Done

jecluis reviewed (changes requested):
LGTM, but I agree with @irq0 that we shouldn't be downloading "random" execution code to our runners.

m-ildefons commented:
I simplified this quite a bit:
- Removed the build matrix. It necessitated to split the tests into groups in advance, which wasn't as simple as just "run all s3tests except these ones that are known to fail". This costs us a bit of parallelism in running tests but also greatly simplifies results collection, which is important.
- Pinned `yq` in place and added checksum checking
- Rebased onto `s3gw` branch
- Quoting and other improvements (typos) to the shell code in the workflow

irq0 reviewed (approved):
lgtm
